### PR TITLE
CAT-1863 Enumerating newly created backgrounds

### PR DESCRIPTION
**Pull request by Achilles-96**
Fixes [CAT-1863](https://jira.catrob.at/browse/CAT-1863)

### CODE REVIEW

**Comment by thmq** [comment]
ok to test

**Comment by Bioxar** [comment]
ok to test

**Comment by thmq** [comment]
ok to test

**Comment by robertpainsi** [comment]
[Utils.getUniqueLookName](https://github.com/Catrobat/Catroid/blob/706599681bd3235f73ced18bf7aaa25703d83440/catroid/src/org/catrobat/catroid/utils/Utils.java#L411) is exactly what you are looking for.

The problem does also address looks from objects and looks chosen from the gallery. I added a [comment](https://jira.catrob.at/browse/CAT-1863?focusedCommentId=17733&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17733) to the issue at this moment.

File: catroid/src/org/catrobat/catroid/ui/fragment/LookFragment.java [hunk 1]
```diff
@@ -614,10 +615,22 @@ public void addLookDrawNewImage() {
 			return;
 		}
 
+		String[] lookNamesPresent = new String[100];
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/ui/fragment/LookFragment.java]: Nowadays you hardly use arrays. You should prefer lists. Take a look at http://stackoverflow.com/a/1589952 for some reasons.

Creating an array with fixed size of `100` is not a good choice.

- Objects can have more than 100 looks
- and you also know how many looks the sprite has. `lookDataList.size()`

-> `new String[lookDataList.size()];`. However, you should always prefer (and like in this case use) some high-level container (like lists).

**Comment by robertpainsi** [comment]
For sounds, the enumration works. However, it's implemented in the SoundController.
https://github.com/Catrobat/Catroid/search?utf8=%E2%9C%93&q=getUniqueSoundName

But I haven't looked into all the details and maybe the LookFragment is a better place than the LookController. You and the person who will review this PR have to take a deeper look at all this me... logic. :worried:

**Comment by thmq** [comment]
ok to test

**Comment by Achilles-96** [comment]
@robertpainsi , I updated PR.

**Comment by robertpainsi** [comment]
Thanks! I :eyes:
I guess you didn't reuse ```Utils.getUniqueLookName(LookData, boolean)``` because the signature doesn't match. Just change it so it does :wink: ```(String, boolean)```. Please create a commit for changing the signature and another one on top of that that actually fixes [CAT-1863](https://jira.catrob.at/browse/CAT-1863).

As you already read in [CAT-1863](https://jira.catrob.at/browse/CAT-1863), the missing enumeration problem does address far more scenarios. It would be awesome if you can take a look at all of them. So that all options how to add a Look (Pocket Paint, Gallery/Photos, ...) work correctly.

I couldn't find any tests addressing the problem in the code base. Please take a look as well. If there are none, you have to write some as well.
@thmq Is there anything to consider for writing tests currently?

Oh, and btw, your implementation ```getFinalLookName``` is better than ```Utils.getUniqueLookName``` :wink:. But please don't change it yet.

File: catroid/src/org/catrobat/catroid/ui/controller/LookController.java [hunk 2]
```diff
@@ -313,6 +313,8 @@ private void copyImageToCatroid(String originalImagePath, Activity activity, Lis
 				imageName = oldFile.getName();
 			}
 
+			imageName = LookFragment.getFinalLookName(imageName,lookDataList);
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/ui/controller/LookController.java]: @thmq 
The [Jenkins build](https://jenkins.catrob.at/job/CatroidCheck/32) is [green](https://jenkins.catrob.at/job/CatroidCheck/32) but running Checkstyle locally will report errors. e.g.

```
[ant:checkstyle] .../catroid/src/org/catrobat/catroid/ui/controller/LookController.java:316:56: warning: ',' is not followed by whitespace.
```

**Comment by Achilles-96** [comment]
@robertpainsi I didn't use `Utils.getUniqueLookName(LookData, boolean)` because from my understanding, it required LookData which represents a fully defined look. But here we just need a name which is used to create a look later. Please correct me if I a wrong. 
One more thing I had in mind is that, as we are not using `Utils.getUniqueLookName(LookData, boolean)` for anything other than name, we can replace it with the function implemented by me.

**Comment by Achilles-96** [comment]
Regarding handling various ways of adding look, I tried to handle adding looks more than once from gallery and also adding multiple looks using camera. Media library shows a dialog that a look already exists and gives an option to rename it. Please tell if I missed any case.

**Comment by robertpainsi** [comment]
ok to test

**Comment by robertpainsi** [comment]
> ... `Utils.getUniqueLookName(LookData, boolean)` ... we can replace it with `getFinalLookName`

Sure :smile_cat:. Just to clarify, keep the `Utils.getUniqueLookName` method but change its signature and implementation. If you do so, please also include at least Sound as well and create a new Pull Request just for your refactoring.

jfi, I will be off until the end of next week. Already procrastinating for my next exam... :worried: So further comments will be delayed as well.

**Comment by Achilles-96** [comment]
I updated PR. Review when you are free.

File: catroid/src/org/catrobat/catroid/ui/controller/LookController.java [hunk 3]
```diff
@@ -313,6 +313,8 @@ private void copyImageToCatroid(String originalImagePath, Activity activity, Lis
 				imageName = oldFile.getName();
 			}
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/ui/controller/LookController.java]: Remove this empty line. Defining the imageName will be a single block.

File: catroid/src/org/catrobat/catroid/utils/Utils.java [hunk 4]
```diff
@@ -408,31 +409,18 @@ private static String searchForNonExistingNfcTagName(String name, int nextNumber
...
-			forBackPack) {
-		String newName;
-		List<LookData> lookDataList;
-		if (forBackPack) {
-			lookDataList = BackPackListManager.getInstance().getAllBackPackedLooks();
-		} else {
-			lookDataList = ProjectManager.getInstance().getCurrentSprite().getLookDataList();
+	public static String getUniqueLookName(String defaultName, List<LookData> presentList) {
+		List<String> lookNamesPresent = new ArrayList<>();
+		for (int i = 0; i < presentList.size(); i++) {
```
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/utils/Utils.java]: Use a for-each loop.

**Comment by robertpainsi** [comment]
Looks great! :+1:
Just one more thing. As mentioned above, please create a separate commit for the refactoring. And on top of it, the commit for the fix.

**Comment by Achilles-96** [comment]
@robertpainsi , Any update on this?

**Comment by robertpainsi** [comment]
Please keep each commit compilable :wink: ([34fd0c2](https://github.com/Catrobat/Catroid/pull/1598/commits/34fd0c2421c260c1a9679793ef83826b66b1ce1f)).
Also rename the second commit message to your PR title. As far as I understand, your fix also fixes the enumeration for looks (not only backgrounds). So just add that to the message as well.

**Comment by Achilles-96** [comment]
Is this fine now.

**Comment by robertpainsi** [comment]
Almost :smile_cat:. If you take a look at the `git log`, your last commit message doesn't meet the guidelines. The PR title would have met. However, if you want to add the word `fix`, use something like `CAT-1863 Fix enumerating for newly created backgrounds and looks` or something similar :wink:.

**Comment by Achilles-96** [comment]
Can this be merged now

**Comment by Achilles-96** [comment]
Any update on this?

**Comment by Achilles-96** [comment]
Should I make a commit over this fixing merge conflict or will you fix it while merging.

**Comment by robertpainsi** [comment]
> Should I make a commit over this fixing merge conflict or will you fix it while merging.

The best way to solve this merge conflict is during rebasing. :wink: It's the same amount of work but much cleaner. So, just checkout your branch and rebase it on the current develop branch. Let me know if you need any more information.

**Comment by Achilles-96** [comment]
I rebased and updated

**Comment by Achilles-96** [comment]
Any update on this?

**Comment by thmq** [comment]
Please squash into one commit, and reopen if still relevant.

**Comment by Achilles-96** [comment]
Sorry for late reply. But I was asked to make it as two commits - one for refactoring and other for the rest by @robertpainsi .

**Pull request closed without merging**